content/doc/developer/handling-requests/json.adoc

```diff
@@ -0,0 +1,244 @@
+
+
+=== Testing with CURL
+
+A `mvn hpi:run` in the plugin should be enough to run it locally.  Assuming that Jenkins is up at http://localhost:8080/jenkins/ , you should have at this point:
+
+[source,bash]
+----
+curl -XGET \
+    -w "\n STATUS:%{http_code}"  \
+    http://localhost:8080/jenkins/custom-api/get-example-param?paramValue=hello
+
+{"message":"I am Jenkins hello"}
+ STATUS:200
+----
+
+=== Example of test with JenkinsRule
+
+[source,java]
+----
+public class JsonAPITest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static String GET_API_URL = "custom-api/get-example-param?paramValue=hello";
+
+    @Test
+    public void testGetJSON() throws Exception {
+
+        // Testing a simple GET that should answer 200 OK and a json
+        JenkinsRule.JSONWebResponse response = j.getJSON(GET_API_URL);
+        assertTrue(response.getContentAsString().contains("I am JenkinsRule hello"));
+        assertEquals(response.getStatusCode(), 200);
+    }
+
+    @Test
+    public void testAdvancedGetJSON() throws Exception {
+        //Testing a GET that requires the user to be authenticated
+        User admin = User.getById("admin", true);
+        MockAuthorizationStrategy auth = new MockAuthorizationStrategy()
+                .grant(Jenkins.ADMINISTER).everywhere().to(admin);
+
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(auth);
+
+        //We need to setup the webclient
+        //By default if the status code is not ok, WebClient throw an exception
+        //Since we want to assert the error status code, we need to set to false.
+        JenkinsRule.WebClient webClient = j.createWebClient();
+        webClient.setThrowExceptionOnFailingStatusCode(false);
+
+        // - simple call without authentication should be forbidden
+        response = j.getJSON(GET_API_URL, webClient);
+        assertEquals(response.getStatusCode(), 403);
+
+        // - same call but authenticated using withBasicApiToken() should be fine
+        response = j.getJSON(GET_API_URL, webClient.withBasicApiToken(admin));
+        assertEquals(response.getStatusCode(), 200);
+    }
+
+----
+
+=== Handling POST with Jenkins
+
+Starting from the class `JsonAPI` provided for GET example, add:
+
+[source,java]
+----
+@POST
+@WebMethod(name = "")
+public JsonHttpResponse create(@JsonBody JSONObject body) {
+    MyJsonObject parsedBody = (MyJsonObject) body.toBean(MyJsonObject.class);
+    //Do any logic required for creation
+    //For the example purpose we just unmarshall and recreate json
+    JSONObject response = JSONObject.fromObject(parsedBody);
+    return new JsonHttpResponse(response, 200);
+}
+
+----
+
+=== Testing with CURL
+
+A `mvn hpi:run` in the plugin should be enough to run it locally.  Assuming that Jenkins is up at http://localhost:8080/jenkins/ , you should have at this point:
+
+Get the crumb.... TBD
```

Comment:
Please propose an example with an API token instead to prevent the need for a CSRF Token in CLI/script like this :-)

Reply:
yes thanks for spotting this

Reply:
done